README.md
# Introduce

Ipio touch driver is implemented by ILI Technology Corp, which is mainly used on its new generation touch ICs, TDDI.

# Support ICs

The following lists which of TDDI IC types supported by the driver.

* ILI9881F
* ILI9881H
* ILI7807G

# Support platform

* Firefly-RK3288 (Rockchip)
* Helio X20 (MT6797)
* Qualcomm
* SPRD

The default in this driver works on Qualcomm platform. If you'd like to port other platforms with this driver, please modify the number shown on below to fit it:

**common.h**
```
/* A platform currently supported by driver */
#define PT_QCOM	1
#define PT_MTK	2
#define PT_SPRD	3
#define TP_PLATFORM PT_QCOM
```

# Interface
This driver supports SPI interface in the version of 1.0.3.0 above. If you'd like to change the interface between I2C and SPI:

```
/* A interface currently supported by driver */
#define I2C_INTERFACE 1
#define SPI_INTERFACE 2
#define INTERFACE SPI_INTERFACE
```

Note that SPI is only used on ILI9881H series.

# Functions

## Gesture

To enable the support of this feature, you can either open it by its node under /proc :

```
echo on > /proc/ilitek/gesture
echo off > /proc/ilitek/gesture
```

or change its variable from the file **config.c** :
```
core_config->isEnableGesture = false;
```

## Check battery status

Some specific cases with power charge may affect on our TDDI IC so that we need to protect it if the event is occurring.

In order to enable this function, the first you need to do is to enable its macro at **common.h**

```
/* Check battery's status in order to avoid some effects from charge. */
#define BATTERY_CHECK
```

Once the function has been built in, you must have to on it by echoing:

```
echo on > /proc/ilitek/check_battery
echo off > /procilitek/check_battery
```

Or you can run it by default by setting its flag as true.

```
ipd->isEnablePollCheckPower = true;
```

## DMA

If your platform needs to use DMA with I2C, you can open its macro from **common.h** :
```
#define ENABLE_DMA
```
Note, it is disabled as default.

## I2C R/W Segment

Some of platforms with i2c bus may have no ability to R/W data with large length at once, so we provide th way of segmental operation.

Before enable its macro, you first need to know how much length your platform could accept. In **core/i2c.c** where you can find its variable:

```
core_i2c->seg_len = 256; // length of segment
```

You can change its value as long as you know the length that your platform can work on.

After that, as mentioned before, you need to enable its macro, which defines at **common.h**, to actually let driver run its function.

```
/* Split the length written to or read from IC via I2C. */
#define I2C_SEGMENT
```

## FW upgrade at boot time
Apart from manual firmware upgrade, we also provide the way of upgrading firmware when system boots initially as long as the verions of firmware

in the IC is different the version that you're going to upgrade.

```
/* Be able to upgrade fw at boot stage */
#define BOOT_FW_UPGRADE
```

## Glove/Proximity/Phone cover

These features need to be opened by the node only.

```
echo enaglove > /proc/ilitek/ioctl  --> enale glove
echo disglove > /proc/ilitek/ioctl  --> disable glove
echo enaprox > /proc/ilitek/ioctl   --> enable proximity
echo disprox > /proc/ilitek/ioctl   --> disable proximity
echo enapcc > /proc/ilitek/ioctl    --> enable phone cover
echo dispcc > /proc/ilitek/ioctl    --> disable phone cover
```

# The metho of debug

To do so, you must firstlly ensure that the version of image is compiled for the debug instead of the user, otherwise it won't allow users to write commands through our device nodes.

## Debug message

It is important to see more details with debug messages if an error happends somehow. To look up them, you can set up debug level via a node called debug_level under /proc.

At the beginning, you should read its node by cat to see what the levels you can choose for.

```
cat /proc/ilitek/debug_level

DEBUG_NONE = 0
DEBUG_IRQ = 1
DEBUG_FINGER_REPORT = 2
DEBUG_FIRMWARE = 4
DEBUG_CONFIG = 8
DEBUG_I2C = 16
DEBUG_BATTERY = 32
DEBUG_MP_TEST = 64
DEBUG_IOCTL = 128
DEBUG_NETLINK = 256
DEBUG_ALL = -1
```

The default level is zero once you get the driver at the first time. Let's say that we want to check out what is going on when the driver is upgrading firmware.

```
echo 4 > /proc/ilitek/debug_level
```

The result will only print the debug message with the process of firmware. Furthermore, you can also add two or three numbers to see multiple debug levels.

```
echo 7 > /proc/ilitek/debug_level
```

In this case the debusg message prints the status of IRQ, FINGER_REPORT and FIRMWARE. Finally, you can definitly see all of them without any thoughts.

```
echo -1 > /proc/ilitek/debug_level
```

## I2C R/W

Sometime it is necessary to see IC's direct output by sending I2C commands in special cases.

We provide three functions to achieve it by echoing a specific device nodev which is created under /proc/ilitek/ioctl.

The command line defines as below :

```
# echo <function>,<length>,<data> > /proc/ilitek/ioctl
```

The comma must be added in each following command to let the driver recognize them.

In the function you can order specific actions to operate I2C such as Write (i2c_w), Read (i2c_r) and Write/Read (i2c_w_r). For instance, if you want to order write command, you can then type this as following:

```
# echo i2c_w,0x3,0x40,0x0,0x1
```

In this case the function is Write, the length is 3, and the i2c command is 0x40, 0x0 and 0x1. Note that it is important to add "0x" with every numbers that you want to cmmand except for the functions.

Additionly, we offer a useful feature to do Write/Read at once: Delay time. In this function it will do Write first and then delay a certain time before reading data through I2C. Note that the default of delay time is 1ms

```
# echo i2c_w_r, <length of write>, <length of read>, <delay time>, <data>
```

# File structure

```
├── common.h
├── core
│   ├── config.c
│   ├── config.h
│   ├── finger_report.c
│   ├── finger_report.h
│   ├── firmware.c
│   ├── firmware.h
│   ├── flash.c
│   ├── flash.h
│   ├── gesture.c
│   ├── gesture.h
│   ├── i2c.c
│   ├── i2c.h
│   ├── Makefile
│   ├── mp_test.c
│   ├── mp_test.h
│   ├── parser.c
│   ├── parser.h
│   ├── protocol.c
│   ├── protocol.h
│   ├── spi.c
│   └── spi.h
├── Makefile
├── platform.c
├── prob.h
├── README.md
└── userspace.c

```

* The concepts of this driver is suppose to hide details, and you should ignore the core functions and follow up the example code **platform.c**
to write your own platform c file or just modify it. 

* If you want to create more nodes on the device driver for user space, just refer to **userspace.c**.

* The directory **Core** includes our touch ic settings, functions and other features. If you'd like to write an application in user space, can just copy the directory to your workspace and call those functions by including header files.

# Porting

## Tell driver what IC types it should support to.

To void any undefined exeception, you should know what types of IC on the platform and what the current type of IC this driver supports for.

```
/* An Touch IC currently supported by driver */
#define CHIP_TYPE_ILI7807	0x7807
#define CHIP_TYPE_ILI9881	0x9881
#define TP_TOUCH_IC		CHIP_TYPE_ILI9881
```
In this case the driver now supports the type of ILI9881.

## DTS example code

```
&i2c1 {
		ts@41 {
			status = "okay";
        	compatible = "tchip,ilitek";
      		reg = <0x41>;
         	touch,irq-gpio = <&gpio8 GPIO_A7 IRQ_TYPE_EDGE_RISING>;
         	touch,reset-gpio = <&gpio8 GPIO_A6 GPIO_ACTIVE_LOW>;
    	};

};
```
In this case the slave address is 0x41, and the name of table calls **tchip,ilitek**. **touch,irq-gpio** and **touch,reset-gpio** represent INT pin and RESET pin separately.


CHANGELOG
v2.0.7.0
[New features]
1. Add reading latch value when doing spi recovery.
2. Add a device node that shows all version information.
3. Add reading FW MP version.
4. Add reading core version with 4 bytes.
5. Add the mp test when the item of peak to peak td fails.
6. Shows variation in csv log.
7. Add reading TVCH and TVCL in short test from ini file.

[Improvements]
1. Optimized the order of command when switches to mp test mode.
2. No longer disable i2cuart flag when it is opened by users.
3. Optimized the flow of i2c/spi esd gesture.
4. Optimized the flow of writting dummy command in suspend.
5. Optimized the display of fw upgrade process.

[Bug fixed]
1. Some variables wasn't freed when driver initializtion failed.
2. The return in strstr() is fixed.
3. Fix the error when reading ini file by request_firmware.
4. Modified the formula of short test.

v2.0.6.0
[New features]
1. Able to adjust spi clock by echoing ioctl (if platforms support).

[Improvements]
1. Remove Noise Diff from MP Test.
2. Adjust the formula of 7807G/R.

[Bus fixed]
1. Fix the issue while getting cdc data in gesture mode.

v2.0.5.0
[New features]
1. Add new mp test item (Noise Diff)
2. Won't go to sleep stop and check busy while system suspending (for core version 1.4.3 above).
3. Add DDI R/W ioctl
4. Now can read iram data in TP modules with flash.
5. Add INI release version and date in CSV log.

[Improvements]
1. Now the results of strings in fw upgrad and mp can print on terminal as well as send to apk.
2. Wrapper strncmp to avoid the failure of comparsion with whole part of string.
3. Read pc count and latch if spi ack error occurs.
4. Rewrite MP LIST and the order to test item that follows with the order of ini section.
5. Add the coordinate of gesture
6. Modify open test c formula for 7807G/R/Q
7. Remove other ILI files in order to reduce the size of driver code.

[Bus fixed]
1. Fix unlock error when geting CDC data.
2. Fix device id error in probe stage.
3. Fix the error of DAC test.
4. Free those allocated memories that wasn't freed.

v2.0.4.0
[New features]
1. Add the option of transferring touch position, which is supported by core version 1.4.3 above.

[Improvements]
1. Add the definitions for various tp modules. Users must implement the function that is called by driver to assign a proper define.
2. Driver sends the results of mp test with the format of strings to apk.

[Bus fixed]
1. Fixed compiled error on rk3288.
2. Wake up spi clock should be called in the first parst of function of spi write.
3. As there is a hw reset action in esd gesture, it won't commad waking spi up in spi write until they finished.
4. The format of sequence of csv log and execution in driver will be the same.

* v2.0.3.1
[Bug fixed]
1. Fix the issue of compiling error in rk3288.
2. Dummy cmd for waking up the clk of spi pll should be placed in very first in write.

* v2.0.3.0
[New features]
1. Add panel/tp info with ioctl for apk

[Improvements]
1. If IC is 7807G from AA to AH version, it must be checking chip id/spi ack while entering/exiting ice mode.
2. Optimized the progress of operating DMA.

[Bus fixed]
1. Fix the issue of not getting versions when auto upgrade is disabled.
2. Skips an interrupt and remove retry when occuring rx lock error.
3. Remove sending dummy cmd in spi read during system suspend.
4. Change the result from fail to pass and the name of csv file as 'warning' if the settings of Golden mode and spec option in ini file aren't matched.

* v2.0.2.0
[New features]
1. Add parsing gesture coordinate.
2. Remove pre command before getting cdcd data if core ver is 1.4.3
above.

[Improvements]
1. Add reading pc latch in pc counter.

[Bus fixed]
1. Add back short test and doez p2p test
2. Fix error of getting panel info from hex.
3. Fix kernel panic while mp test failed in timeout of int.
4. Correct the address of esd gesture with iram.

* v2.0.1.4
[New features]

[Improvements]
1. Chech size and crc while parsing ILI file.
2. Allocated larger buffers at probe in advance
3. Optimized the handle of exepction erros with fw upgrade.
3. Add duplex in spi split.

[Bus fixed]
1. Fix compling error with qcom.
2. Fix error of loading gesture code in last version.

* v2.0.1.3
[New features]
1.  ilitek_fw.c 拆分成 ilitek_flash.c 和 ilitek_hostdl.c

[Improvements]
1. FW更新後，從HEX讀取 TP info (SPI/I2C預設打開)
2. 優化proximity判斷
3. 沒有golden時，在做完MP後會產出”mp_warning”的csv
4. 更新 open sp測試的公式
5. 調整top half裡判斷順序
6. SPI要load gesture code時，加入判斷是否動態load code
7. 減少解析ini的時間
8. SPI esd gesture位址從0x25FF8，改成0x40054

[Bus fixed]
1. 重新加入gesture debug
2. 修正在debug mode下，Android系統未出現小白點的問題
3. 在做 esd gesture時，load 完 ap code 會 sleep 50ms (搭配info_from_hex和core ver 1.4.1)
4. 修正 checkpatch.pl 錯誤
5. 修正讀取報點時，若出現錯誤(非 recovery)會沒有返回的問題

* v2.0.1.2
1.	在MP測試裡，增加錯誤返回值
2.	修正open test c 測試異常
3.	修正DeltaC Record修正，合入OPPO功能需求驗證完成
4.	在喚醒SPI PLL Clock時，若SPI為15Mhz，需寫10 bytes來符合Timing，所以9 bytes 改為10 bytes
5.	修正會誤啟動gesture recovery動作
6.	重新改寫 dump iram data function
7.	優化SPI通信時間  #67  (需開啟SPI_DMA_TRANSFER_SPLIT，但X20是需要切割，故只驗報點功能)
8.	重新加入pm_stay_awake，某次改動被拿掉了。

* v2.0.1.1
1.	加入OPPO需求功能 (gesture failure reason, demo debug …etc) (驗證中)
2.	加入 wake up (dummy) command，用來在暗屏時，喚醒SPI PLL Clock
3.	優化TP/FW Mode state machine 寫法
	* TP mode只會有3種模式 : AP, Gesture, MP
	* 其餘像是Debug/Demo Debug等等，皆為不同的資料格式
	* 一個AP mode會有多種不同資格上報格式
	* 資料格式長度、算法集中放在ilitek_set_tp_data_len處理
4.	加入高通的DRM_MSM休眠/喚醒事件機制
5.	加入讓客戶可以從DDI直接調用TP喚醒的callback function
6.	加入保存Hex資料機制，避免系統獲取不到system的partition導致固件版本回退
7.	修正copy_to_user和copy_from_user的返回判斷值
8.	修改sprint-->snprintf , strcmp-->strncmp
9.	加入mp.ini沒有benchmark的話，會產出warning檔名的csv file
10.	加入PM stay awake/relax，避免系統(或PM)進了深層睡眠導致手勢失敗
11.	修正MP的CSV格式 (Test Studio可直接取Golden)
12.	修正Whole Chip Reset的資料和地址錯誤
13.	修正Cove version取4 bytes的錯誤 (只需取3 bytes)
14.	加入SPI R/W調試的echo (不進入spi ice)
15.	加入在手勢模式下，可直接收取Debug資料
16.	修正做host download時，資料長度不足2k時會計算錯誤的問題
17.	SPI Recovery / Gesture Recovery 改成直接調用方式，不再透過workqueue來排程
18.	新增MP的PIN測項
19.	加入所有function的返回值判斷，避免被Coverity tool找到錯誤

* v2.0.1.0

common.h
/*
 * ILITEK Touch IC driver
 *
 * Copyright (C) 2011 ILI Technology Corporation.
 *
 * Author: Dicky Chiang <dicky_chiang@ilitek.com>
 * Based on TDD v7.0 implemented by Mstar & ILITEK
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 */

#ifndef __COMMON_H
#define __COMMON_H

#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/init.h>
#include <linux/slab.h>
#include <linux/fs.h>
#include <linux/types.h>
#include <linux/kernel.h>
#include <linux/errno.h>
#include <linux/i2c.h>
#include <linux/input.h>
#include <linux/platform_device.h>
#include <linux/kobject.h>
#include <linux/interrupt.h>
#include <linux/delay.h>
#include <linux/version.h>
#include <linux/regulator/consumer.h>
#include <linux/power_supply.h>
#include <linux/fs.h>
#include <asm/uaccess.h>

#include <linux/cdev.h>
#include <linux/device.h>
#include <linux/proc_fs.h>
#include <linux/string.h>
#include <linux/ctype.h>

#include <linux/netlink.h>
#include <linux/skbuff.h>
#include <linux/socket.h>
#include <net/sock.h>

#include <linux/sched.h>
#include <linux/kthread.h>
#include <linux/wait.h>
#include <linux/time.h>

#include <linux/namei.h>
#include <linux/vmalloc.h>
#include <linux/timer.h>
#include <linux/jiffies.h>
#include <linux/module.h>
#include <linux/dma-mapping.h>

#include <linux/gpio.h>
#include <linux/spi/spi.h>
#include <linux/rtc.h>
#include <linux/syscalls.h>
#include <linux/security.h>
#include <linux/mount.h>

#ifdef CONFIG_OF
#include <linux/of_address.h>
#include <linux/of_device.h>
#include <linux/of_irq.h>
#include <linux/of.h>
#include <linux/of_gpio.h>
#endif

#ifdef CONFIG_FB
#include <linux/notifier.h>
#include <linux/fb.h>
#else
#include <linux/earlysuspend.h>
#endif



/* A interface currently supported by driver */
#define I2C_INTERFACE 1
#define SPI_INTERFACE 2
#define INTERFACE I2C_INTERFACE

/* Driver version */
#define DRIVER_VERSION	"1.0.4.0"

/* Driver core type */
#define CORE_TYPE_B		0x00
#define CORE_TYPE_E		0x03

/* Protocol version */
#define PROTOCOL_MAJOR		0x5
#define PROTOCOL_MID		0x4
#define PROTOCOL_MINOR		0x0

/*  Debug messages */
#ifdef BIT
#undef BIT
#endif
#define BIT(x)	(1 << (x))

enum {
	DEBUG_NONE = 0,
	DEBUG_IRQ = BIT(0),
	DEBUG_FINGER_REPORT = BIT(1),
	DEBUG_FIRMWARE = BIT(2),
	DEBUG_CONFIG = BIT(3),
	DEBUG_I2C = BIT(4),
	DEBUG_BATTERY = BIT(5),
	DEBUG_MP_TEST = BIT(6),
	DEBUG_IOCTL = BIT(7),
	DEBUG_NETLINK = BIT(8),
	DEBUG_PARSER = BIT(9),
	DEBUG_GESTURE = BIT(10),
	DEBUG_ALL = ~0,
};

#define ipio_info(fmt, arg...)	\
	pr_info("ILITEK: (%s, %d): " fmt, __func__, __LINE__, ##arg);

#define ipio_err(fmt, arg...)	\
	pr_err("ILITEK: (%s, %d): " fmt, __func__, __LINE__, ##arg);

#define ipio_debug(level, fmt, arg...)									\
	do {																\
		if (level & ipio_debug_level)									\
		pr_info("ILITEK: (%s, %d): " fmt, __func__, __LINE__, ##arg);	\
	} while (0)

/* Reset methods */
enum {
	SW_RST = 0,
	HW_RST,
	HOST_DOWNLOAD_RST,
	HOST_DOWNLOAD_BOOT_RST
};

/* Distributed to all core functions */
extern uint32_t ipio_debug_level;
extern uint32_t ipio_chip_list[2];

/* Macros */
#define CHECK_EQUAL(X, Y) 	((X == Y) ? 0 : -1)
#define ERR_ALLOC_MEM(X)	((IS_ERR(X) || X == NULL) ? 1 : 0)
#define MIN(a,b) (((a)<(b))?(a):(b))
#define MAX(a,b) (((a)>(b))?(a):(b))
#define USEC	1
#define MSEC	(USEC * 1000)

/* The definition for firmware upgrade */
#define MAX_HEX_FILE_SIZE			(160*1024)
#define MAX_FLASH_FIRMWARE_SIZE		(256*1024)
#define MAX_IRAM_FIRMWARE_SIZE		(60*1024)
#define ILI_FILE_HEADER				64
#define MAX_AP_FIRMWARE_SIZE		(64*1024)
#define MAX_DLM_FIRMWARE_SIZE		(8*1024)
#define MAX_MP_FIRMWARE_SIZE		(64*1024)
#define MAX_GESTURE_FIRMWARE_SIZE	(8*1024)
#define MAX_TUNING_FIRMWARE_SIZE	(4*1024)
#define MAX_DDI_FIRMWARE_SIZE		(4*1024)

#define DLM_START_ADDRESS           0x20610
#define DLM_HEX_ADDRESS             0x10000
#define MP_HEX_ADDRESS              0x13000
#define SPI_UPGRADE_LEN				2048
#define FW_BLOCK_INFO_NUM			6
#define UPDATE_RETRY_COUNT			3

#define AP_BLOCK_NUM					1
#define DATA_BLOCK_NUM					2
#define TUNING_BLOCK_NUM				3
#define GESTURE_BLOCK_NUM				4
#define MP_BLOCK_NUM					5
#define DDI_BLOCK_NUM					6

#define ILITEK_I2C_ADDR			0x41

/*
 * ILI9881 Series
 */
enum ili9881_types {
	TYPE_F = 0x0F,
	TYPE_H = 0x11
};

#define ILI9881_ICE_MODE_ADDR	0x181062
#define ILI9881_PID_ADDR		0x4009C
#define ILI9881_OTP_ID_ADDR		0x400A0
#define ILI9881_ANA_ID_ADDR		0x400A4
#define ILI9881_PC_COUNTER_ADDR 0x44008
#define ILI9881_WDT_ADDR		0x5100C
#define ILI9881_CHIP_RESET_ADDR 0x40050

/*
 * ILI7807 Series
 */
enum ili7807_types {
	TYPE_G = 0x10,
};

#define ILI7807_ICE_MODE_ADDR	0x181062
#define ILI7807_PID_ADDR		0x4009C
#define ILI7807_OTP_ID_ADDR		0x400A0
#define ILI7807_ANA_ID_ADDR		0x400A4
#define ILI7807_PC_COUNTER_ADDR 0x44008
#define ILI7807_WDT_ADDR		0x5100C
#define ILI7807_CHIP_RESET_ADDR 0x40050

/*
 * Other settings
 */
#define CSV_LCM_ON_PATH     	"/sdcard/ilitek_mp_lcm_on_log"
#define CSV_LCM_OFF_PATH	"/sdcard/ilitek_mp_lcm_off_log"
#define INI_NAME_PATH		"/sdcard/mp.ini"
#define UPDATE_FW_PATH		"/sdcard/ILITEK_FW"
#define POWER_STATUS_PATH 	"/sys/class/power_supply/battery/status"
#define CHECK_BATTERY_TIME  2000
#define CHECK_ESD_TIME		4000
#define VDD_VOLTAGE			1800000
#define VDD_I2C_VOLTAGE		1800000

 /* define the width and heigth of a screen. */
#define TOUCH_SCREEN_X_MIN 0
#define TOUCH_SCREEN_Y_MIN 0
#define TOUCH_SCREEN_X_MAX 720
#define TOUCH_SCREEN_Y_MAX 1280

/* define the range on panel */
#define TPD_HEIGHT 2048
#define TPD_WIDTH 2048

/* How many numbers of touch are supported by IC. */
#define MAX_TOUCH_NUM	10

/* It's only for spi interface used to download data to iram */
#if (INTERFACE == SPI_INTERFACE)
#define HOST_DOWNLOAD
#endif

/* Linux multiple touch protocol, either B type or A type. */
#define MT_B_TYPE

/* Enable the support of regulator power. */
//#define REGULATOR_POWER_ON

/* Either an interrupt event handled by kthread or work queue. */
#define USE_KTHREAD

/* Enable DMA with I2C. */
//#define I2C_DMA

/* Split the length written to or read from IC via I2C. */
//#define I2C_SEGMENT

/* Be able to upgrade fw at boot stage */
//#define BOOT_FW_UPGRADE

/* Read fw file by request_firmware */
//#define BOOT_FW_UPGRADE_READ_HEX

#ifdef BOOT_FW_UPGRADE_READ_HEX
#define BOOT_FW_HEX_NAME "ilitek_fw.hex"
#define BOOT_UPDATE_FW_DELAY_TIME 10000
#endif

/* Enable gesture function */
//#define GESTURE_ENABLE

/* Check battery's status in order to avoid some effects from charge. */
//#define BATTERY_CHECK

/* Check whether the IC is damaged by ESD */
//#define ESD_CHECK

static inline void ipio_kfree(void **mem) {
	if (*mem != NULL) {
		kfree(*mem);
		*mem = NULL;
	}
}

static inline void ipio_vfree(void **mem) {
	if (*mem != NULL) {
		vfree(*mem);
		*mem = NULL;
	}
}

static inline void *ipio_memcpy(void *dest, const void *src, size_t n, size_t dest_size) {
    if (n > dest_size)
         n = dest_size;

    return memcpy(dest, src, n);
}

extern int katoi(char *string);
extern int str2hex(char *str);
#endif /* __COMMON_H */
